src/app/clusters/device-energy-management-server/DeviceEnergyManagementCluster.cpp

```diff
@@ -14,438 +16,386 @@
-
-using namespace chip;
-using namespace chip::app;
-using namespace chip::app::Clusters;
-using namespace chip::app::Clusters::DeviceEnergyManagement;
-using namespace chip::app::Clusters::DeviceEnergyManagement::Attributes;
+#include <clusters/DeviceEnergyManagement/Structs.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <protocols/interaction_model/StatusCode.h>
+#include <system/SystemClock.h>
 
 using chip::Protocols::InteractionModel::Status;
-
 namespace chip {
 namespace app {
 namespace Clusters {
-namespace DeviceEnergyManagement {
 
-CHIP_ERROR Instance::Init()
+namespace {
+// Helper function to add a status response when an error occurs during command handling
+DataModel::ActionReturnStatus AddResponseOnError(const DataModel::InvokeRequest & request, CommandHandler * handler,
+                                                 DataModel::ActionReturnStatus status)
 {
-    ReturnErrorOnFailure(CommandHandlerInterfaceRegistry::Instance().RegisterCommandHandler(this));
-    VerifyOrReturnError(AttributeAccessInterfaceRegistry::Instance().Register(this), CHIP_ERROR_INCORRECT_STATE);
-
-    return CHIP_NO_ERROR;
+    if (status.IsError())
+    {
+        handler->AddStatus(request.path, status.GetStatusCode());
+    }
+    return status;
 }
 
-void Instance::Shutdown()
+bool IsWithinRange(const int64_t power, const uint32_t duration,
+                   const DeviceEnergyManagement::Structs::PowerAdjustCapabilityStruct::Type & powerAdjustmentCapability)
 {
-    TEMPORARY_RETURN_IGNORED CommandHandlerInterfaceRegistry::Instance().UnregisterCommandHandler(this);
-    AttributeAccessInterfaceRegistry::Instance().Unregister(this);
+    if (powerAdjustmentCapability.powerAdjustCapability.IsNull())
+    {
+        return false;
+    }
+
+    for (const auto & pas : powerAdjustmentCapability.powerAdjustCapability.Value())
+    {
+        if ((power >= pas.minPower) && (duration >= pas.minDuration) && (power <= pas.maxPower) && (duration <= pas.maxDuration))
+        {
+            return true;
+        }
+    }
+    return false;
 }
 
-bool Instance::HasFeature(Feature aFeature) const
+// Helper function to validate that the ESA is in the expected state before processing a command
+DataModel::ActionReturnStatus ValidateESAState(const DataModel::InvokeRequest & request, CommandHandler * handler,
+                                               DeviceEnergyManagement::Delegate * delegate,
+                                               DeviceEnergyManagement::ESAStateEnum expectedState)
 {
-    return mFeature.Has(aFeature);
+    DeviceEnergyManagement::ESAStateEnum currentState = delegate->GetESAState();
+    if (currentState != expectedState)
+    {
+        ChipLogError(Zcl, "DEM: ESAState mismatch - expected %d, got %d", static_cast<int>(expectedState),
+                     static_cast<int>(currentState));
+        return AddResponseOnError(request, handler, Status::InvalidInState);
```

Comment:
```suggestion
        return Status::InvalidInState;
```

is probably what you want and this applies throughout. See my other comment.